### PR TITLE
Attempting to test automatically the `_keys_to_ignore`.

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

This adds a new part of the `tied_weights` test that aims at detecting automatically
when `_keys_to_ignore` is incorrectly set.

`_keys_to_ignore` aims to ignore weights that are supposed to be tied in the 
final model, meaning it's OK if the parameter is missing from the on-disk weights.
The weights are really empty during the load, but they end up being tied afterwards
so we should ignore them during the load if they are missing.

The test also aims to detect `_keys_to_ignore` that might have been set but 
could be misleading because the parameters are actually NOT tied anymore.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/bert/modeling_bert.py [hunk 1]
```diff
@@ -1153,7 +1155,7 @@ def forward(
 class BertLMHeadModel(BertPreTrainedModel):
 
     _keys_to_ignore_on_load_unexpected = [r"pooler"]
-    _keys_to_ignore_on_load_missing = [r"position_ids", r"predictions.decoder.bias"]
+    _keys_to_ignore_on_load_missing = [r"position_ids", r"predictions.decoder.bias", r"cls.predictions.decoder.weight"]
```
Comment (ydshieh) [inline src/transformers/models/bert/modeling_bert.py:1158]: Let me review this slowly 🙏  and verify a few things. But do you think `r"predictions.decoder.bias"` is a mistake and should be `r"cls.predictions.decoder.bias"` ?

I am going to check myself anyway.

[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/models/bert/modeling_bert.py:1158]: A re match is done, so while the exact name is indeed `r"cls.predictions.decoder.bias"`, this works. But would be great to fix just in case one day a weight named `predictions.decoder.bias` that should not be ignore appears ;-)

Comment (Narsil) [reply src/transformers/models/bert/modeling_bert.py:1158]: Maybe. I'm not sure but the test does yell if we're hiding a valid key (we I don't try to yell when we have an unused `_keys_to_ignore`. Want me to try and add it to the test ?

**Review by sgugger — COMMENTED** [review]
Interesting way of testing those attributes!

File: tests/test_modeling_common.py [hunk 2]
```diff
@@ -1538,6 +1538,36 @@ def check_same_values(layer_1, layer_2):
             # # Check that the embedding layer and decoding layer are the same in size and in value
             # self.assertTrue(model.transformer.wte.weight.shape, model.lm_head.weight.shape)
             # self.assertTrue(check_same_values(model.transformer.wte, model.lm_head))
+            with tempfile.TemporaryDirectory() as d:
```
[the remark this answers is not in the document]
Comment (Narsil) [reply tests/test_modeling_common.py:1541]: Moving to its own test + comment on how it works.

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -2422,7 +2422,8 @@ def _fix_key(key):
 
         if remove_prefix_from_model:
             expected_keys_not_prefixed = [s for s in expected_keys if not s.startswith(prefix)]
-            expected_keys = [".".join(s.split(".")[1:]) if s.startswith(prefix) else s for s in expected_keys]
+            _prefix = f"{prefix}."
+            expected_keys = [".".join(s.split(".")[1:]) if s.startswith(_prefix) else s for s in expected_keys]
```
Comment (Narsil) [inline src/transformers/modeling_utils.py:2426]: This would fail for `esm`/`esmfold` since `esm` is a prefix for `esmfold` it would remove `emsfold.` prefix instead of the intended `esm.`

This was silent before, but with the new test addition the following modification would fail to get `key[-1]` when trying to identify missing keys.

File: tests/test_modeling_common.py [hunk 4]
```diff
@@ -1485,11 +1485,24 @@ def test_correct_missing_keys(self):
             base_model_prefix = model.base_model_prefix
 
             if hasattr(model, base_model_prefix):
+
```
Comment (Narsil) [inline tests/test_modeling_common.py:1488]: This is an important new modification.

This test tries to verify that we're correctly warning when we're saving the "core" of a model, and failing to save the head parameters, and that we yell when loading the head from this.

This test is great, and different from the newly added one.
But, since we fixed a lot of `Decoder` head which are merely a copy of the embeddings matrix, then they **are** a head, but they DO NOT have new extra parameters. hence this test would fail.

There is `self.test_missing_keys` flag to deactivate that particular test on the model testing. But it would deactivate ALL potential heads (as far as I understand), potentially opening gaps in our testing.

So I modified this test to try and detect those edge cases. I don't like putting logic like that in test as it makes reasoning hard but I failed to see another way.

This test verifies that the head does have extra parameters before actually doing the verification.

File: tests/test_modeling_common.py [hunk 5]
```diff
@@ -1539,6 +1552,49 @@ def check_same_values(layer_1, layer_2):
...
+                    ipdb.set_trace()
+
+                self.assertEqual(
+                    extra_missing,
+                    set(),
+                    f"This model {model_class.__name__} might be missing some `keys_to_ignore`: {extra_missing} when"
+                    " tying its weights",
+                )
+
+                # self.assertEqual(
```
Comment (Narsil) [inline tests/test_modeling_common.py:1591]: I deactivated this part because `token_type_ids` and `position_ids` would yell pretty much everywhere.

I have no idea of the perfect solution for this.
I **think** `position_ids` should be a non-parameter if it's not, or a parameter if it is learned or something. 
Given how many models it affect we could add an escape hatch for just these peculiar names.

But I think flagging **actually** too inclusive `_keys_to_ignore` can provide good value, especially since we're using regexp so it's easy to make a mistake, or modify the model later in such a way that it doesn't reflect properly anymore in those keys.

**Comment by ydshieh** [comment]
@Narsil 

I am not able to reproduce for the `splinter` test you mentioned above. Could you provide more information about your environment and how you launched the test?

File: tests/test_modeling_common.py [hunk 6]
```diff
@@ -1539,6 +1552,44 @@ def check_same_values(layer_1, layer_2):
             # self.assertTrue(model.transformer.wte.weight.shape, model.lm_head.weight.shape)
             # self.assertTrue(check_same_values(model.transformer.wte, model.lm_head))
 
+    def test_tied_model_weights_key_ignore(self):
```
Comment (ydshieh) [inline tests/test_modeling_common.py:1555]: I would suggest to make the name more generic (i.e. not using `tied_model` in the name)

The check in this test is not really specific to tied weights, although they are one of the main source of missing keys.

`test_model_weights_key_ignore` + a comment to explain its objective would be fine.

File: tests/test_modeling_common.py [hunk 7]
```diff
@@ -1539,6 +1552,44 @@ def check_same_values(layer_1, layer_2):
...
+        config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
+        for model_class in self.all_model_classes:
+            model_tied = model_class(config)
+            with tempfile.TemporaryDirectory() as d:
+                model_tied.save_pretrained(d)
+
+                # We are nuking ALL weights on file, so every parameter should
+                # yell on load. We're going to detect if we yell too much, or too little.
+                with open(os.path.join(d, "pytorch_model.bin"), "wb") as f:
+                    torch.save({}, f)
```
Comment (ydshieh) [inline tests/test_modeling_common.py:1565]: A comment here would be nice, as this is unusual.

File: tests/test_modeling_common.py [hunk 8]
```diff
@@ -1539,6 +1552,44 @@ def check_same_values(layer_1, layer_2):
...
+                model_reloaded, infos = model_class.from_pretrained(d, output_loading_info=True)
+
+                prefix = f"{model_reloaded.base_model_prefix}."
+                params = dict(model_reloaded.named_parameters())
+                params.update(dict(model_reloaded.named_buffers()))
```
Comment (ydshieh) [inline tests/test_modeling_common.py:1570]: You mentioned on DM 

> They are listed as two different things during from_pretrained but only appear once when doing model.named_parameters()

I haven't 100% understand it (despite got some rough feeling about it). Could you explain

- why not using `model_reloaded.state_dict()` -- (from your word, I guess we will get 2 things, and it will match the missing keys, which will pass the test when the situation should fail)
- Give an example, say what happens to bert or gpt2

In any case, a comment (inside the code) should be added here regarding the usage of `named_parameters` and `named_parameters` 🙏 .

**Review by ydshieh — COMMENTED** [review]
Just a few comments - only for the test `test_tied_model_weights_key_ignore` so far.

File: tests/test_modeling_common.py (hunk 7 above)
Comment (Narsil) [reply tests/test_modeling_common.py:1565]: I gave you the wrong link (on an old commit, the comment is actually there).

Is it readable ?

Comment (ydshieh) [reply tests/test_modeling_common.py:1565]: Oh, yes! Sorry,  I actually didn't understand that `We are nuking ALL weights on file, so every parameter should ...` which is quite clear already.

File: tests/test_modeling_common.py [hunk 9]
```diff
@@ -1539,6 +1552,55 @@ def check_same_values(layer_1, layer_2):
...
+
+                missing_keys = set(infos["missing_keys"])
+
+                extra_missing = missing_keys - param_names
+                # missed_missing = param_names - missing_keys
+
+                self.assertEqual(
+                    extra_missing,
+                    set(),
+                    f"This model {model_class.__name__} might be missing some `keys_to_ignore`: {extra_missing} when"
```
Comment (ydshieh) [inline tests/test_modeling_common.py:1593]: Forgot to mention, we could remove `when tying its weights` here.

File: src/transformers/modeling_utils.py (hunk 3 above)
Comment (sgugger) [reply src/transformers/modeling_utils.py:2426]: Seems like the right fix, thanks!

File: src/transformers/modeling_utils.py [hunk 10]
```diff
@@ -2641,7 +2642,9 @@ def retrieve_modules_from_names(self, names, add_prefix=False, remove_prefix=Fal
 
         # torch.nn.ParameterList is a special case where two parameter keywords
         # are appended to the module name, *e.g.* bert.special_embeddings.0
-        module_keys = module_keys.union(set([".".join(key.split(".")[:-2]) for key in names if key[-1].isdigit()]))
+        module_keys = module_keys.union(
+            set([".".join(key.split(".")[:-2]) for key in names if key and key[-1].isdigit()])
+        )
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:2647]: Why the change here? Can we revert if it's just styling?

File: src/transformers/models/bart/modeling_bart.py [hunk 11]
```diff
@@ -1451,6 +1458,11 @@ def _reorder_cache(past, beam_idx):
     BART_START_DOCSTRING,
 )
 class BartForSequenceClassification(BartPretrainedModel):
+    _keys_to_ignore_on_load_missing = [
+        "encoder.embed_tokens.weight",
+        "decoder.embed_tokens.weight",
+    ]
```
Comment (sgugger) [inline src/transformers/models/bart/modeling_bart.py:1464]:
```suggestion
    _keys_to_ignore_on_load_missing = ["encoder.embed_tokens.weight", "decoder.embed_tokens.weight"]
```

File: src/transformers/models/bart/modeling_bart.py [hunk 12]
```diff
@@ -1578,6 +1590,11 @@ def forward(
     BART_START_DOCSTRING,
 )
 class BartForQuestionAnswering(BartPretrainedModel):
+    _keys_to_ignore_on_load_missing = [
+        "encoder.embed_tokens.weight",
+        "decoder.embed_tokens.weight",
+    ]
```
Comment (sgugger) [inline src/transformers/models/bart/modeling_bart.py:1596]:
```suggestion
    _keys_to_ignore_on_load_missing = ["encoder.embed_tokens.weight", "decoder.embed_tokens.weight"]
```

File: src/transformers/models/bigbird_pegasus/modeling_bigbird_pegasus.py [hunk 13]
```diff
@@ -2651,6 +2658,11 @@ def _reorder_cache(past, beam_idx):
 )
 # Copied from transformers.models.bart.modeling_bart.BartForSequenceClassification with Bart->BigBirdPegasus, BART->BIGBIRD_PEGASUS
 class BigBirdPegasusForSequenceClassification(BigBirdPegasusPreTrainedModel):
+    _keys_to_ignore_on_load_missing = [
+        "encoder.embed_tokens.weight",
+        "decoder.embed_tokens.weight",
+    ]
```
Comment (sgugger) [inline src/transformers/models/bigbird_pegasus/modeling_bigbird_pegasus.py:2664]:
```suggestion
    _keys_to_ignore_on_load_missing = ["encoder.embed_tokens.weight", "decoder.embed_tokens.weight"]
```

File: src/transformers/models/bigbird_pegasus/modeling_bigbird_pegasus.py [hunk 14]
```diff
@@ -2779,6 +2791,11 @@ def forward(
 )
 # Copied from transformers.models.bart.modeling_bart.BartForQuestionAnswering with Bart->BigBirdPegasus, BART->BIGBIRD_PEGASUS
 class BigBirdPegasusForQuestionAnswering(BigBirdPegasusPreTrainedModel):
+    _keys_to_ignore_on_load_missing = [
+        "encoder.embed_tokens.weight",
+        "decoder.embed_tokens.weight",
+    ]
```
Comment (sgugger) [inline src/transformers/models/bigbird_pegasus/modeling_bigbird_pegasus.py:2797]:
```suggestion
    _keys_to_ignore_on_load_missing = ["encoder.embed_tokens.weight", "decoder.embed_tokens.weight"]
```

File: src/transformers/models/layoutlmv2/image_processing_layoutlmv2.py [hunk 15]
```diff
@@ -187,7 +187,6 @@ def preprocess(
         tesseract_config: Optional[str] = None,
         return_tensors: Optional[Union[str, TensorType]] = None,
         data_format: ChannelDimension = ChannelDimension.FIRST,
-        **kwargs,
```
Comment (sgugger) [inline src/transformers/models/layoutlmv2/image_processing_layoutlmv2.py:190]: Doesn't seem related to this PR, why is it here?

File: src/transformers/models/layoutlmv3/image_processing_layoutlmv3.py [hunk 16]
```diff
@@ -261,7 +261,6 @@ def preprocess(
         tesseract_config: Optional[str] = None,
         return_tensors: Optional[Union[str, TensorType]] = None,
         data_format: ChannelDimension = ChannelDimension.FIRST,
-        **kwargs,
```
Comment (sgugger) [inline src/transformers/models/layoutlmv3/image_processing_layoutlmv3.py:264]: Same here

File: src/transformers/models/longt5/modeling_longt5.py [hunk 17]
```diff
@@ -2137,7 +2137,7 @@ def _reorder_cache(self, past, beam_idx):
     LONGT5_START_DOCSTRING,
 )
 class LongT5EncoderModel(LongT5PreTrainedModel):
-    authorized_missing_keys = [
+    _keys_to_ignore_on_load_missing = [
         r"encoder.embed_tokens.weight",
     ]
```
Comment (sgugger) [inline src/transformers/models/longt5/modeling_longt5.py:2142]: Let's make this one-line

File: src/transformers/models/mvp/modeling_mvp.py [hunk 18]
```diff
@@ -1291,7 +1291,14 @@ def custom_forward(*inputs):
...
 class MvpModel(MvpPreTrainedModel):
-    _keys_to_ignore_on_load_unexpected = [r"final_logits_bias", r"lm_head.weight"]
+    _keys_to_ignore_on_load_unexpected = [
+        r"final_logits_bias",
+        r"lm_head.weight",
+    ]
+    _keys_to_ignore_on_load_missing = [
+        "encoder.embed_tokens.weight",
+        "decoder.embed_tokens.weight",
+    ]
```
Comment (sgugger) [inline src/transformers/models/mvp/modeling_mvp.py:1301]: Boths can fit in one line.

File: src/transformers/models/prophetnet/modeling_prophetnet.py [hunk 19]
```diff
@@ -859,11 +859,7 @@ def forward(
...
 
-        assert list(hidden_states.size()) == [
-            batch_size,
-            ngram_sequence_length,
-            hidden_size,
-        ], (
+        assert list(hidden_states.size()) == [batch_size, ngram_sequence_length, hidden_size], (
             f"`hidden_states` should be of shape {batch_size, ngram_sequence_length, hidden_size}, but is of shape"
             f" {hidden_states.shape}"
         )
```
Comment (sgugger) [inline src/transformers/models/prophetnet/modeling_prophetnet.py:865]: Seems unrelated, and if we are going to change it, let's make it a real test and an exception raised :-p

File: src/transformers/models/xlm_prophetnet/modeling_xlm_prophetnet.py [hunk 20]
```diff
@@ -876,11 +876,7 @@ def forward(
     ):
         batch_size, ngram_sequence_length, hidden_size = hidden_states.size()
 
-        assert list(hidden_states.size()) == [
-            batch_size,
-            ngram_sequence_length,
-            hidden_size,
-        ], (
+        assert list(hidden_states.size()) == [batch_size, ngram_sequence_length, hidden_size], (
```
Comment (sgugger) [inline src/transformers/models/xlm_prophetnet/modeling_xlm_prophetnet.py:879]: Same as before.

File: tests/test_modeling_common.py (hunk 4 above)
Comment (sgugger) [reply tests/test_modeling_common.py:1488]: Make sure you account in the test for upcoming models which have weights with the same name in the model with head and the base model (Niels is adding one in #19981)

File: tests/test_modeling_common.py (hunk 8 above)
Comment (ydshieh) [reply tests/test_modeling_common.py:1581]: As long as using named_parameters/named_buffer works, I wouldn't request to this fancier check :-)

**Review by sgugger — COMMENTED** [review]
Amazing work! Some styling comments and parts that seem orthogonal to this PR.

File: src/transformers/modeling_utils.py (hunk 10 above)
Comment (Narsil) [reply src/transformers/modeling_utils.py:2647]: It's not just styling I'm adding `if key and key[-1]` vs `if key[-1]` it should prevent failing even if there is a bad prefix.

I still could revert though. I left it to prevent further failure, but we're already in a bad state if this fails so...

File: src/transformers/models/layoutlmv2/image_processing_layoutlmv2.py (hunk 15 above)
Comment (Narsil) [reply src/transformers/models/layoutlmv2/image_processing_layoutlmv2.py:190]: I have no idea.

Comment (Narsil) [reply src/transformers/models/layoutlmv2/image_processing_layoutlmv2.py:190]: I think I just bulked reverted some changes I had there at some point. I think I did `git checkout main src/tr... ` maybe my main wasn't up-to-date.

File: src/transformers/models/prophetnet/modeling_prophetnet.py (hunk 19 above)
Comment (Narsil) [reply src/transformers/models/prophetnet/modeling_prophetnet.py:865]: I'm pretty sure it's my black that does this. However it should be 22.3 Like `transformers expect. Not sure where the diff comes from.

And `fixup` doesn't revert those :(

File: tests/test_modeling_common.py (hunk 6 above)
Comment (Narsil) [reply tests/test_modeling_common.py:1555]: In my mind it is only about tied weights (because it's the only mechanism I know that can create the weight sharing).

Happy to update the name

File: src/transformers/models/mvp/modeling_mvp.py (hunk 18 above)
Comment (Narsil) [reply src/transformers/models/mvp/modeling_mvp.py:1301]: Done.

File: src/transformers/models/xlm_prophetnet/modeling_xlm_prophetnet.py (hunk 20 above)
Comment (Narsil) [reply src/transformers/models/xlm_prophetnet/modeling_xlm_prophetnet.py:879]: Done.

File: src/transformers/modeling_utils.py (hunk 10 above)
Comment (Narsil) [reply src/transformers/modeling_utils.py:2647]: It was `""` so `len(key) > 0` it is.

File: src/transformers/models/layoutlmv3/image_processing_layoutlmv3.py (hunk 16 above)
Comment (Narsil) [reply src/transformers/models/layoutlmv3/image_processing_layoutlmv3.py]: Reverted those

File: src/transformers/models/xlm_prophetnet/modeling_xlm_prophetnet.py (hunk 20 above)
Comment (Narsil) [reply src/transformers/models/xlm_prophetnet/modeling_xlm_prophetnet.py:879]: Black is not happy about this the old code it forces me to use this new layout.

I tried to revert this.

File: tests/test_modeling_common.py (hunk 4 above)
Comment (Narsil) [reply tests/test_modeling_common.py:1471]: I don't see which weights are sharing names.

Seems relatively regular to me: https://github.com/huggingface/transformers/pull/19981/files#diff-a2ff7478a3e58383ae065b28d9e5e4e1e1bace51b32bc91b6eb086fabb0bd702R448-R470

**Review by LysandreJik — APPROVED** [review]
Impressive work! The changes in common look good to me. 
The tests look quite robust as well.

File: src/transformers/modeling_utils.py (hunk 3 above)
Comment (Narsil) [reply src/transformers/modeling_utils.py:2426]: Ok, there was a second place to update the fix to make the `splinter` test pass.

Comment (Narsil) [reply src/transformers/modeling_utils.py:2426]: https://github.com/huggingface/transformers/pull/20042/files#diff-6b72b98c4c2dcfc6cc606843917733f5d858374fbc22a735ff483bbc0c1e63eaR2655

**Review by sgugger — APPROVED** [review]

**Comment by younesbelkada** [comment]
Hey! 
I am not sure if it is because of this PR but loading NLLB (that is affected by this PR) now gives:
```
│ /home/younes_huggingface_co/debug_issues/code/transformers/src/transformers/modeling_utils.py:24 │
│ 59 in _load_pretrained_model                                                                     │
│                                                                                                  │
│   2456 │   │   │   for key in missing_keys:                                                      │
│   2457 │   │   │   │   if key.startswith(prefix):                                                │
│   2458 │   │   │   │   │   key = ".".join(key.split(".")[1:])                                    │
│ ❱ 2459 │   │   │   │   param = model_state_dict[key]                                             │
│   2460 │   │   │   │   if param.device == torch.device("meta"):                                  │
│   2461 │   │   │   │   │   if not load_in_8bit:                                                  │
│   2462 │   │   │   │   │   │   set_module_tensor_to_device(model, key, "cpu", torch.empty(*para  │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
KeyError: 'encoder.embed_positions.weights'
```
Here is the snippet to reproduce the error:
```
from transformers import AutoModelForSeq2SeqLM, AutoTokenizer

src_lang = "eng_Latn"
tgt_lang = "spa_Latn"

tokenizer = AutoTokenizer.from_pretrained("facebook/nllb-200-distilled-600M", src_lang=src_lang)
model = AutoModelForSeq2SeqLM.from_pretrained("facebook/nllb-200-distilled-600M",
                                              device_map= "auto")
```
I did not followed entirely this PR but I will dig into that now and see what exactly caused the issue 💪

File: src/transformers/models/m2m_100/modeling_m2m_100.py [hunk 21]
```diff
@@ -1244,12 +1246,8 @@ class M2M100ForConditionalGeneration(M2M100PreTrainedModel):
         r"encoder.version",
         r"decoder.version",
         r"lm_head.weight",
-        r"model.encoder.embed_positions.weights",
```
Comment (younesbelkada) [inline src/transformers/models/m2m_100/modeling_m2m_100.py:1247]: I realized these are removed, is this intended (I see that it's one of the only model where some keys are removed)? Maybe it has been done by mistake but I am not sure

Comment (Narsil) [reply src/transformers/models/m2m_100/modeling_m2m_100.py:1247]: Seems I forgot to re-add this. 

This was removed because of the test which is now commented which tries to remove `_key_to_ignore` which are masking real weights.

This SinusoidalPositionEmbeddings, seems to be a buffer and not a parameter, but something is wrong somewhere.
Since it's a nn.Module, maybe it still sees  `weights` as being a Parameter somehow, idk.

We can safely revert this part anyway.

Comment (younesbelkada) [reply src/transformers/models/m2m_100/modeling_m2m_100.py:1247]: Thanks a lot for the explanation ! 🚀
